extensions/amp-analytics/0.1/requests.js

```diff
@@ -0,0 +1,231 @@
+   * @private
+   */
+  trigger_() {
+    if (!this.schedule_) {
+      this.schedule_ = this.win.setTimeout(() => {
+        this.fire_();
+      }, this.maxDelay_ * 1000);
+    }
+  }
+
+  /**
+   * Send out request.
+   * @private
+   */
+  fire_() {
+    //TODO: need to preconnect to request url at the right time
+    this.baseUrlPromise_.then(request => {
+      return Promise.all(this.extraUrlParamsPromise_).then(paramStrs => {
+        filterSplice(paramStrs, item => {return !!item;});
+        const extraUrlParamsStr = paramStrs.join('&');
+        if (request.indexOf('${extraUrlParams}') >= 0) {
+          request = request.replace('${extraUrlParams}', extraUrlParamsStr);
+        } else {
+          request = appendEncodedParamStringToUrl(request, extraUrlParamsStr);
+        }
+        return request;
+      });
+    }).then(request => {
+      const lastTrigger = /** @type {!JsonObject} */ (this.lastTrigger_);
+      this.handler_(request, lastTrigger);
+      this.reset_();
+    });
+  }
+
+  /**
+   * Reset batching status
+   * @private
+   */
+  reset_() {
+    this.baseUrlPromise_ = null;
+    this.extraUrlParamsPromise_ = [];
+    this.schedule_ = null;
+    this.lastTrigger_ = null;
+  }
+
+  /**
+   * Function that handler extraUrlParams from config and trigger.
+   * @param {!JsonObject} configParams
+   * @param {!JsonObject} triggerParams
+   * @param {!./variables.ExpansionOptions} expansionOption
+   * @return {!Promise<string>}
+   * @private
+   */
+  expandExtraUrlParams_(configParams, triggerParams, expansionOption) {
+    const requestPromises = [];
+    const params = map();
+    // Add any given extraUrlParams as query string param
+    if (configParams || triggerParams) {
+      Object.assign(params, configParams, triggerParams);
+      for (const k in params) {
+        if (typeof params[k] == 'string') {
+          requestPromises.push(
+              this.variableService_.expandTemplate(params[k], expansionOption)
+                  .then(value => { params[k] = value; }));
+        }
+      }
+    }
+    return Promise.all(requestPromises).then(() => {
+      return this.getExtraUrlParamsString_(params);
+    });
+  }
+
+  //TODO: Need function to handler batchSegment as well
+
+  /**
+   * Handle the params map and form the final extraUrlParams string
+   * @param {!Object} params
+   * @return {string}
+   */
+  getExtraUrlParamsString_(params) {
+    const s = [];
+    for (const k in params) {
+      const v = params[k];
+      if (v == null) {
+        continue;
+      } else {
+        const sv = this.variableService_.encodeVars(v, k);
+        s.push(`${encodeURIComponent(k)}=${sv}`);
+      }
+    }
+    return s.join('&');
+  }
+}
+
+/**
+ * Expand config's request to object
+ * @param {!JsonObject} config
+ */
+export function expandConfigRequest(config) {
+  if (!config['requests-v2']) {
+    return config;
+  }
+  for (const k in config['requests-v2']) {
+    if (hasOwn(config['requests-v2'], k)) {
```

Comment:
Isn't it dangerous to expand inline? At least we need to make sure we do not expand twice?

Reply:
hmmm I think it's ok. We only expand the config to an object if it is not, should never expand twice.